### ISSUE TITLE
how do you login / fix age restriction?

### ISSUE THREAD

**Issue by C0DK** [pytube/pytube#84]
I am getting the following error:
raise AgeRestricted("Age restricted video. Unable to download "
pytube.exceptions.AgeRestricted: Age restricted video. Unable to download without being signed in.

and i obviously just have to sign in; but how do i sign in via pytube? have i missed a part of the docs or is it not implemented?

Thanks!

**Comment by nficano** [comment]
@C0DK Perhaps I should be raising a not implemented error here -- I kicked around the idea of allowing users to configure their login, but my fear is it would make the whole fetching process brittle, having to obtain and pass around the session. Do you find you encounter this on a lot of videos?

**Comment by C0DK** [comment]
maybe 1/10th of the videos i want to download. i could go around and use my win machine for these rare cases but i'd be nice if it simply logged in IF it got a age-restriction. (then it wont waste time everytime - only when needed)

but what ever floats your boat. i'm just happy to be able to use the programs as it is!

**Comment by nficano** [comment]
I looked into youtube-dl (which is what I use as my measuring stick for pytube), they already do this, so I'm going to try and find a few hours this weekend to get this in.

**Comment by C0DK** [comment]
awesome! :+1:

**Comment by dyeray** [comment]
Is it really needed to log in to obtain a link? Any age restricted video can be played without being logged in by replacing "watch?v=" by "embed/", so there should be a way to download it without logging in.

**Comment by C0DK** [comment]
dueray great point actually. however i think that @nficano is using YT API to some extend (haven't really looked at the source files..)

**Comment by nficano** [comment]
I've tried this once before to no avail, let me try again later tonight

**Comment by dyeray** [comment]
I have taken a look into youtube.com/embed/ and into your code. Maybe what I have found can be useful. What you call json_object is not present on the embed page, but something very similar is obtained from the embed page using ajax from https://www.youtube.com/get_video_info (doing a Google search, I can see this url is widely used to obtain youtube download links). The content is url encoded, but as I said it is very similar to what you have on json_object. There is also a 'url_encoded_fmt_stream_map' that contains the url encoded stream map. The urls to the video are on the 'url' key.

These links to the videos work, but the 'signature=' is not there. I am not sure what is the purpose of the signature, though. Also I think this method doesn't work with videos that can't be embedded or at least I saw some error with a Vevo video.

**Comment by ShaneGlee** [comment]
@nficano did you get anywhere with the age restriction issue?

**Comment by AttitudeMonger** [comment]
Is there any fix to this problem yet? 

Surprisingly, when I run my program from my Mac command prompt, it works fine. However, when I have a cloud Linux VM installed from AWS and I try it there, I get this error. Somehow Mac command prompt uses the cookie information stored in Safari or Chrome I guess (I should verify by logging out of Google in those browsers and trying the program again).

**Comment by TobyMellor** [comment]
+1

**Comment by wgaylord** [comment]
Any progress on this. I was temped to use a login and pass a CookieJar to all the urlopen functions after I log in. 

https://docs.python.org/2/library/cookielib.html#examples

**Comment by bjogden** [comment]
+1

**Comment by ADB991** [comment]
If you implement the login functionality, maybe we can start downloading YouTube Red videos that one has paid.

**Comment by ADiegoCAlonso** [comment]
+1

**Comment by nficano** [comment]
On the roadmap

**Comment by Jestonyeta** [comment]
update ?

**Comment by nficano** [comment]
Couple weeks out, finishing up documentation, then playlists, then this, if by all means anyone wants to help contribute, it would be appreciated.

**Comment by grking8** [comment]
PR https://github.com/nficano/pytube/pull/179 submitted for this

**Comment by CTimmerman** [comment]
Same appears to happen now.
YouTube URL: https://www.youtube.com/watch?v=eKyxJDUiYPg
Dead or Alive Xtreme 2 Intro HD
ERROR:eKyxJDUiYPg is age restricted, and can't be accessed without logging in.
Traceback (most recent call last):
  File "C:\Users\C\Documents\code\pytube\youtube-dl.py", line 65, in <module>
    download_youtube_stream(input("YouTube URL: "))
  File "C:\Users\C\Documents\code\pytube\youtube-dl.py", line 42, in download_youtube_stream
    streams = youtube.streams
  File "C:\Users\C\AppData\Local\Programs\Python\Python38\lib\site-packages\pytube\__main__.py", line 292, in streams
    return StreamQuery(self.fmt_streams)
  File "C:\Users\C\AppData\Local\Programs\Python\Python38\lib\site-packages\pytube\__main__.py", line 172, in fmt_streams
    stream_manifest = extract.apply_descrambler(self.streaming_data)
  File "C:\Users\C\AppData\Local\Programs\Python\Python38\lib\site-packages\pytube\__main__.py", line 156, in streaming_data
    self.bypass_age_gate()
  File "C:\Users\C\AppData\Local\Programs\Python\Python38\lib\site-packages\pytube\__main__.py", line 260, in bypass_age_gate
    raise exceptions.AgeRestrictedError(self.video_id)
pytube.exceptions.AgeRestrictedError: eKyxJDUiYPg is age restricted, and can't be accessed without logging in.

**Comment by 0m3gaV2** [comment]
I have use_oauth set to true and i logged in but I'm still getting
  pytube.exceptions.AgeRestrictedError:  is age restricted, and can't be accessed without logging in.

Fixed it by setting allow_oauth_cache to false to log in again to another google account that doesn't have YouTube TV accessed to it yet, but it seems like setting it back to true after logging in brings back the error mentioned before, so you have to log in for each video

**Comment by heksbeks** [comment]
Once in a while I get false pytube.exceptions.AgeRestrictedError exceptions on videos that are not age restricted. On the second try then it mostly works.

**Comment by heksbeks** [comment]
This video raises the exceptions although it is not age restricted:
https://www.youtube.com/watch?v=ax4H1ikrJn8

**Comment by kaarlych** [comment]
Hello, anyone know how to fix the issue?

**Comment by aardabayram** [comment]
Hi, any updates on the issue?

**Comment by JonJon565** [comment]
[youtube.com/watch?v=tzD9OxAHtzU](https://www.youtube.com/watch?v=tzD9OxAHtzU)
also raises the exception how too fix this?

**Comment by Kyojur0** [comment]
been some time any solution ? without logging in

**Comment by heksbeks** [comment]
This might be a way to solve this: take a tool that is able to download the videos that fail with pytube.exceptions.AgeRestrictedError, for example https://www.4kdownload.com or others, use wireshark to see how it's done

**Comment by kpollackhinds** [comment]
> 
The first solution on this stack overflow works: https://stackoverflow.com/questions/75791765/how-to-download-videos-that-require-age-verification-with-pytube

**Comment by kpollackhinds** [comment]
> [youtube.com/watch?v=tzD9OxAHtzU](https://www.youtube.com/watch?v=tzD9OxAHtzU) also raises the exception how too fix this?

https://stackoverflow.com/questions/75791765/how-to-download-videos-that-require-age-verification-with-pytube

**Comment by kpollackhinds** [comment]
> been some time any solution ? without logging in

First solution here works:
https://stackoverflow.com/questions/75791765/how-to-download-videos-that-require-age-verification-with-pytube

**Comment by kpollackhinds** [comment]
> Hello, anyone know how to fix the issue?

First solution here works:

https://stackoverflow.com/questions/75791765/how-to-download-videos-that-require-age-verification-with-pytube

**Comment by eboriley** [comment]
Thanks 

> > Hello, anyone know how to fix the issue?
> 
> First solution here works:
> 
> https://stackoverflow.com/questions/75791765/how-to-download-videos-that-require-age-verification-with-pytube

Thanks! Your solution works

**Comment by listmen001** [comment]
https://stackoverflow.com/questions/75791765/how-to-download-videos-that-require-age-verification-with-pytube
dosn't work.
ERROR:root:lis3l0IszA0 is age restricted, and can't be accessed without logging in.

**Comment by jacobsilverman** [comment]
reopen this issue, I'm also still facing this issue after following the given solutions

**Comment by girmay** [comment]
This works:  yt = YouTube(url,use_oauth=True, allow_oauth_cache=True)
You will have to standby for onetime auth just once. pytube will generate app-specific password that you will put in https://www.google.com/device

**Comment by GabrielLins64** [comment]
> This works: yt = YouTube(url,use_oauth=True, allow_oauth_cache=True) You will have to standby for onetime auth just once. pytube will generate app-specific password that you will put in https://www.google.com/device

I tried and got the following error:

`regex_search: could not find match for (?:v=|\/)([0-9A-Za-z_-]{11}).*`

**Comment by jacobsilverman** [comment]
I gave up trying t o use pytube and just swapped over to `import yt_dlp as youtube_dl` instead. 

Download the GET cookies.txt LOCALLY chrome extension then you can use that txt file in the yt_dlp configuration so the cookiesfile sets the login.

**Comment by treedbk** [comment]
While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.

That fixed the issue for me.

**Comment by djkemmet** [comment]
> While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.
> 
> That fixed the issue for me.

Thank you for this, this solved my issue as well! Warm regards.

**Comment by theseriff** [comment]
> While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.
> 
> That fixed the issue for me.

Thank you very much! it's work

**Comment by kingmacth** [comment]
> While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.
> 
> That fixed the issue for me.

After modifying it in your way, it still prompts "pytube.exceptions.AgeRestrictedError: Z1uJl8T0ix4 is age restricted, and can't be accessed without logging in."
"

**Comment by g-i-o-r-g-i-o** [comment]
still buggy

**Comment by r00tm4ch1n3** [comment]
> While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.
> 
> That fixed the issue for me.

Works for me. Thank you.

**Comment by jilp-ai** [comment]
> > While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.
> > That fixed the issue for me.
> 
> Works for me. Thank you.

Worked for me as well. Thank you!

**Comment by loopkinng** [comment]
> While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.
> 
> That fixed the issue for me.

Thank you, sir. Worked for me as well. 
> I was downloading a YT Short. However, downloading long videos was working well.

**Comment by todowede** [comment]
> While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.
> 
> That fixed the issue for me.

Worked for me too. Thanks

**Comment by shouge** [comment]
> While using pytube version 15.0.0 I went into the pytube folder, opened `__main__.py`, and changed line 253 from `client='ANDROID_EMBED',` to `client='ANDROID',`.
> 
> That fixed the issue for me.

Thank you!

**Comment by valentimaraujo** [comment]
Based on the solution [@GTS](https://stackoverflow.com/a/76780768/4895962), I solved my problem in the following way, without needing to modify the source code:

```
# pytube==15.0.0
from pytube import YouTube
from pytube.innertube import _default_clients

_default_clients["ANDROID_MUSIC"] = _default_clients["ANDROID"]
extractYB = YouTube(source_url)

video_stream = extractYB.streams.get_highest_resolution()
destination_path = './video/youtube'
video_stream.download(output_path=destination_path)
video_size_bytes = video_stream.filesize
video_duration_seconds = extractYB.length
```

**Comment by tewnut** [comment]
> Based on the solution [@GTS](https://stackoverflow.com/a/76780768/4895962), I solved my problem in the following way, without needing to modify the source code:
> 
> ```
> # pytube==15.0.0
> from pytube import YouTube
> from pytube.innertube import _default_clients
> 
> _default_clients["ANDROID_MUSIC"] = _default_clients["ANDROID"]
> extractYB = YouTube(source_url)
> 
> video_stream = extractYB.streams.get_highest_resolution()
> destination_path = './video/youtube'
> video_stream.download(output_path=destination_path)
> video_size_bytes = video_stream.filesize
> video_duration_seconds = extractYB.length
> ```

For my case, the ANDROID as default client leads to error 400. So changing to this works fine:

`_default_clients["ANDROID_MUSIC"] = _default_clients["ANDROID_CREATOR"]`

**Comment by Tboiii-123** [comment]
Hi guys, pls who has gotten a solution for the age restriction in downloading Youtube videos

**Comment by Tboiii-123** [comment]
import yt_dlp

# List of YouTube video URLs you want to download
video_urls = [
    'https://youtube.com/shorts/nAD-1ffz6yk?si=gq9e4t2BiO-p9uXl'
    
    #youtube link here
]

# Replace with your desired output directory

output_path = 'C:/Users/HP/Downloads/'

# Options for yt_dlp (YouTube Downloader)
ydl_opts = {
    'format': 'best',  # Select the best available format
    'outtmpl': output_path + '%(title)s.%(ext)s',  # Output file template
    'quiet': True,  # Suppress output messages
}

# Lists to store successful and failed download URLs
failed_downloads = []
successful_downloads = []

# Initialize YouTube Downloader with the provided options
with yt_dlp.YoutubeDL(ydl_opts) as ydl:
    for url in video_urls:
        try:
            ydl.download([url])  # Download the video
            print("Downloaded:", url)
            successful_downloads.append(url)
        except Exception as e:
            print(f"Failed to download {url}: {e}")
            failed_downloads.append(url)
            
# Print results
print("\n" * 4)
print("Failed downloads:", failed_downloads)
print("Successful downloads:", successful_downloads)
Best solution have found so far try it out guys
